truffle/src/main/java/org/jruby/truffle/runtime/subsystems/ObjectSpaceManager.java

```diff
@@ -55,6 +58,10 @@ public void clearFinalizers() {
     private final ReferenceQueue<DynamicObject> finalizerQueue = new ReferenceQueue<>();
     private DynamicObject finalizerThread;
 
+    private final CyclicAssumption tracingAssumption = new CyclicAssumption("objspace-tracing");
+    private int tracingAssumptionActivations = 0;
+    private boolean tracingPaused = false;
```

Comment:
Probably `ThreadLocal<Boolean>` would work, but of course there is some overhead.
And AtomicInteger for `tracingAssumptionActivations`.